[LoopVectorize] Improve algorithm for hoisting runtime checks

david-arm opened the pull request:
When attempting to hoist runtime checks out of a loop we currently avoid creating pointer diff checks and prefer to do expanded range checks instead. This gives us the opportunity to hoist runtime checks out of a loop, since these checks are loop invariant. However, in some cases the pointer diff checks would also be loop invariant and so will naturally get hoisted. Therefore, since diff checks are cheaper so we should prefer to use those instead.

david-arm committed:
[LoopVectorize] Improve algorithm for hoisting runtime checks

When attempting to hoist runtime checks out of a loop we currently
avoid creating pointer diff checks and prefer to do expanded range
checks instead. This gives us the opportunity to hoist runtime
checks out of a loop, since these checks are loop invariant. However,
in some cases the pointer diff checks would also be loop invariant
and so will naturally get hoisted. Therefore, since diff checks are
cheaper so we should prefer to use those instead.

nikic commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -346,7 +346,10 @@ void RuntimePointerChecking::tryToCreateDiffCheck(
     if (StartARLoop == SinkStartAR->getLoop() &&
-        StartARLoop == InnerLoop->getParentLoop()) {
+        StartARLoop == InnerLoop->getParentLoop() &&
+        !SE->isKnownPredicate(ICmpInst::ICMP_EQ,
+                              SrcStartAR->getStepRecurrence(*SE),
+                              SinkStartAR->getStepRecurrence(*SE))) {
```
```suggestion
        SrcStartAR->getStepRecurrence(*SE) != SinkStartAR->getStepRecurrence(*SE)) {
```
Don't use isKnownPredicate() for ICMP_EQ.

nikic replied:
Yeah, SCEVs are uniqued, so a SCEV with the same args will have the same pointer. (Of course, it's possible that expressions are equal in ways that SCEV doesn't understand, but we'd consider that a canonicalization failure.)

david-arm committed:
Don't use isKnownPredicate for checking equality of SCEVs

david-arm committed:
Try attaching data layout to test

The new test I added was borrowed from runtime-checks-difference.ll
and seems to be failing with buildkite, but I cannot reproduce the
failure locally.

Just as an experiment I am also copying the datalayout from
runtime-checks-difference.ll to see if that fixes it.

david-arm commented:
The clang formatting errors have nothing to do with my patch - they all seem to complain about existing code.

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -346,7 +346,9 @@ void RuntimePointerChecking::tryToCreateDiffCheck(
     if (StartARLoop == SinkStartAR->getLoop() &&
-        StartARLoop == InnerLoop->getParentLoop()) {
+        StartARLoop == InnerLoop->getParentLoop() &&
+        SrcStartAR->getStepRecurrence(*SE) !=
```
Would be good to add a comment here explaining the check

david-arm replied:
Hi @fhahn, sorry I only just saw this now! I find with github it's really easy to miss review comments compared to Phabricator...

fhahn replied:
Yeah the UI does unfortunately sometimes hide some comments :(

fhahn reviewed (commented):
The test is also failing when I check out the branch for the PR.

It looks like the branch is based on d4cec1ce7313608670d5a97d7c9fcccad68ac659, which doesn't have the diff check improvements I think, which is probably why it is failing.

david-arm commented:
> The test is also failing when I check out the branch for the PR.
> 
> It looks like the branch is based on [d4cec1c](https://github.com/llvm/llvm-project/commit/d4cec1ce7313608670d5a97d7c9fcccad68ac659), which doesn't have the diff check improvements I think, which is probably why it is failing.

OK thanks I'll try rebasing!

david-arm committed:
Merge branch 'llvm:main' into diff_checks

david-arm committed:
Add comments explaining why we sometimes prefer diff checks

paulwalker-arm commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -347,7 +347,12 @@ void RuntimePointerChecking::tryToCreateDiffCheck(
-        StartARLoop == InnerLoop->getParentLoop()) {
+        StartARLoop == InnerLoop->getParentLoop() &&
+        // If the diff check would already be loop invariant (due to the
+        // recurrences being the same), then we should still prefer the diff
```
Perhaps "...,then we prefer to keep the diff check because they are cheaper."

fhahn reviewed (approved):
LGTM, thanks for the update!

david-arm committed:
Rewrite diff check comment
